pkg/elle/core/graph.go

```diff
@@ -19,75 +19,263 @@ type DirectedGraph struct {
 func (g *DirectedGraph) Out(v Vertex) []Vertex {
-	panic("impl me")
+	_, ok := g.Outs[v]
+	if !ok {
+		return []Vertex{}
+	}
+
+	vertices := []Vertex{}
+
+	for key := range g.Outs[v] {
+		vertices = append(vertices, key)
+	}
+	return vertices
 }
 
 // Edges returns the edge between two vertices
 func (g *DirectedGraph) Edges(a, b Vertex) []Edge {
-	panic("impl me")
+	_, oka := g.Outs[a]
+	_, okb := g.Ins[b]
+	if !oka || !okb {
+		return []Edge{}
+	}
+
+	_, ok := g.Outs[a][b]
+	if !ok {
+		return []Edge{}
+	}
+
+	edges := []Edge{}
+	for _, rel := range g.Outs[a][b] {
+		edges = append(edges, Edge{
+			From:  a,
+			To:    b,
+			Value: rel,
+		})
+	}
+	return edges
 }
 
 // Link links two vertices relationship
-func (g *DirectedGraph) Link(v Vertex, succ []Vertex, rel Rel) {
-	panic("impl me")
+func (g *DirectedGraph) Link(v Vertex, succ Vertex, rel Rel) {
+	_, ok := g.Outs[v]
+	if !ok {
+		g.Outs[v] = make(map[Vertex][]Rel)
+	}
+	_, ok = g.Outs[succ]
+	if !ok {
+		g.Outs[succ] = make(map[Vertex][]Rel)
+	}
+
+	_, ok = g.Ins[v]
+	if !ok {
+		g.Ins[v] = []Vertex{}
+	}
+	_, ok = g.Ins[succ]
+	if !ok {
+		g.Ins[succ] = []Vertex{}
+	}
+
+	haveRel := false
+	for _, item := range g.Outs[v][succ] {
+		if item == rel {
+			haveRel = true
+			break
+		}
+	}
+	if haveRel == false {
+		g.Outs[v][succ] = append(g.Outs[v][succ], rel)
+	}
+
+	haveV := false
+	for _, item := range g.Ins[succ] {
+		if item == v {
+			haveV = true
+			break
+		}
+	}
+	if haveV == false {
+		g.Ins[succ] = append(g.Ins[succ], v)
+	}
 }
 
 // LinkToAll links x to all ys
 func (g *DirectedGraph) LinkToAll(x Vertex, ys []Vertex, rel Rel) {
-	panic("impl me")
+	for _, y := range ys {
+		g.Link(x, y, rel)
+	}
 }
 
 // LinkAllTo links all xs to y
-func (g *DirectedGraph) LinkAllTo(xs []Vertex, y Vertex) {
-	panic("impl me")
+func (g *DirectedGraph) LinkAllTo(xs []Vertex, y Vertex, rel Rel) {
+	for _, x := range xs {
+		g.Link(x, y, rel)
+	}
 }
 
 func (g *DirectedGraph) UnLink(a, b Vertex) {
-	panic("impl me")
+	delete(g.Outs[a], b)
+
+	for id , vertex := range g.Ins[b] {
+		if vertex == a {
+			g.Ins[b] = append(g.Ins[b][:id], g.Ins[b][id+1:]...)
+			break
+		}
+	}
 }
 
 // Fork implements `forked` semantics on DirectedGraph
 func (g *DirectedGraph) Fork() *DirectedGraph {
 	return deepcopy.Copy(g).(*DirectedGraph)
 }
-
-// keepEdgeValues transforms a graph by applying a function (f edge-value) to each edge in the
-//  graph. Where the function returns `nil`, removes that edge altogether
-func (g *DirectedGraph) keepEdgeValues(f func(interface{}) interface{}) *DirectedGraph {
-	panic("impl me")
-}
-
-// RemoveRelationship removes the given relationship from it
-func (g *DirectedGraph) RemoveRelationship(rel Rel) *DirectedGraph {
-	panic("impl me")
+// IntersectionRel returns the union of res1 and res2
+func IntersectionRel(rels1 []Rel, rels2 []Rel) []Rel{
+	rel := []Rel{}
+	for _, rel1 := range rels1 {
+		both := false
+		for _, rel2 := range rels2{
+			if rel1 == rel2 {
+				both = true
```

Comment:
```suggestion
				rel = append(rel, rel1)
```

don't need the `both` variable.